File: src/main/java/org/radarcns/util/StreamUtil.java
package org.radarcns.util;

import org.apache.kafka.streams.KeyValue;

import java.util.function.BiFunction;
import java.util.function.BiPredicate;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

public final class StreamUtil {
    private StreamUtil() {
        // utility class
    }

    public static <K, V> Predicate<KeyValue<K, V>> test(BiPredicate<? super K, ? super V> bip) {
        return entry -> bip.test(entry.key, entry.value);
    }

    public static <K, V, R> Function<KeyValue<K, V>, R> apply(
            BiFunction<? super K, ? super V, R> bif) {
        return entry -> bif.apply(entry.key, entry.value);
    }

    public static <K, V> Function<KeyValue<K, V>, K> first() {
        return e -> e.key;
    }

    public static <K, V> Function<KeyValue<K, V>, V> second() {
        return e -> e.value;
    }

    @FunctionalInterface
    public interface StreamSupplier<T> {
        Stream<T> get();

        default StreamSupplier<T> concat(StreamSupplier<? extends T> other) {
            return () -> Stream.concat(get(), other.get());
        }

        static <T> StreamSupplier<T> supply(StreamSupplier<T> supplier) {
            return supplier;
        }
    }
}

File: src/main/kotlin/org/radarbase/kotlin/util/StreamUtil.kt
package org.radarbase.kotlin.util

import org.apache.kafka.streams.KeyValue
import java.util.stream.Stream

object StreamUtil {
    fun <K, V> test(bip: (K, V) -> Boolean): (KeyValue<K, V>) -> Boolean =
        { entry -> bip(entry.key, entry.value) }

    fun <K, V, R> apply(bif: (K, V) -> R): (KeyValue<K, V>) -> R =
        { entry -> bif(entry.key, entry.value) }

    fun <K, V> first(): (KeyValue<K, V>) -> K = { it.key }
    fun <K, V> second(): (KeyValue<K, V>) -> V = { it.value }

    fun interface StreamSupplier<T> {
        fun get(): Stream<T>

        fun concat(other: StreamSupplier<out T>): StreamSupplier<T> =
            StreamSupplier { Stream.concat(get(), other.get()) }

        companion object {
            fun <T> supply(supplier: StreamSupplier<T>): StreamSupplier<T> = supplier
        }
    }
}
